README.md
zmakebas
--------

Description
-----------

zmakebas converts a Spectrum Basic program written as a text file into
an actual speccy Basic file (as a .TAP file, or optionally a raw
headerless file).

Using zmakebas rather than (say) writing the Basic in an emulator
means you can write using a nicer editor, and can use tools which work
on text files, etc. Also, with the `-l` option you can write without
line numbers, using labels in their place where necessary.

The program was originally intended to be used simply to make little
loader programs, so they wouldn't have to be sourceless binaries.
However, I went to a fair amount of effort to make sure it'd work for
bigger, more serious programs too, so you can also use it for that
kind of thing.


Installation
------------

While I think zmakebas should be fairly portable, I've only ever
compiled it on Linux and MS-DOS. Things should be ok for Linux as-is -
just do `make` then (as root) `make install`.


Miscellaneous
-------------

See the man page for details of how to use zmakebas.

There's a short demo file showing the use of escape sequences for UDGs
and block graphics characters etc. here, named `demo.bas`. You can use
`zmakebas demo.bas` to generate an `out.tap` file from it. A version
using labels is in `demolbl.bas`, which can be converted with
`zmakebas -l demolbl.bas`.


Contacting me
-------------

You can email me at russell.marks@ntlworld.com.


Share and enjoy! 

-Rus.


ChangeLog
2005-07-29 John Elliott  <seasip.webmaster@gmail.com>

	* Version 1.3.

	* Added ability to output files in +3DOS format


2004-07-23  Russell Marks  <russell.marks@ntlworld.com>

	* Version 1.2.

2004-05-01  Russell Marks  <russell.marks@ntlworld.com>

	* Fixed ignoring of escape sequences in REM statements. Thanks
	again to Matthew Westcott.

	* zmakebas.c: added support for embedding literal eight-bit
	character codes into the output, for e.g. colour control codes.
	Thanks to Matthew Westcott for the patch.

2000-11-03  Russell Marks  <russell.marks@ntlworld.com>

	* Version 1.1.

	* zmakebas.1: minor fixes.

	* Cleaned things up a little, made installation saner, added
	uninstall target, and updated contact details.


demolbl.bas
# (for Emacs) -*- indented-text -*-

# this is a label-using version of `demo.bas', which shows how much
# nicer it is not to have to deal with line numbers. :-)


rem zmakebas demo

gosub @init
gosub @header
gosub @udgdem
gosub @blockdem
stop


@init:
for f=usr "a"+7 to usr "u"+7 step 8
poke f,255
next f
let is48=1

# init all attrs just in case
border 7:paper 7:ink 7:bright 0:flash 0:cls

# check for 48k speccy or 48k mode. This is a pretty nasty way to do
# it, but seems to be the most sane way (from Basic at least).
print "\t"
if screen$(0,0)="S" then let is48=0
ink 0:print at 0,0;
return


@header:
print ink 5;"\..\..\..\..\..\..\..\..\..\..\..\..\..\..\..\..";\
            "\..\..\..\..\..\..\..\..\..\..\..\..\..\..\.."
print paper 5;"  Non-ASCII chars in zmakebas  "
print ink 5;"\''\''\''\''\''\''\''\''\''\''\''\''\''\''\''\''";\
	    "\''\''\''\''\''\''\''\''\''\''\''\''\''\''\''"
print
return


@udgdem:
print "Here are the UDGs:"''
print ink 1;"\a\b\c\d\e\f\g\h\i\j\k\l\m\n\o\p\q\r\s";
if is48 then print ink 1;"\t\u";
print ''"(They should be underlined.)"
return


@blockdem:
#              01234567890123456789012345678901
print at 11,0;"The block graphics, first as"'\
	      "listed by a for..next loop, then"'\
	      "via zmakebas's escape sequences:"
ink 7
print at 15,0;
for f=128 to 143:print chr$(f);" ";:next f:print ''
print at 17,0;"\   \ ' \'  \'' \ . \ : \'. \': ";\
	      "\.  \.' \:  \:' \.. \.: \:. \::"
# draw boxes around them to make it look less confusing
ink 1
for y=0 to 1
for x=0 to 15
plot x*16,55-y*16:draw 7,0:draw 0,-7:draw -7,0:draw 0,7
next x
next y
ink 0
print at 20,0;"And finally here's the copyright"'\
	      "symbol (";ink 1;"\*";ink 0;\
	      ") and pound sign (";ink 1;"`";ink 0;")."
return


demo.bas
# (for Emacs) -*- indented-text -*-

# this file `demo.bas' demonstrates the features of zmakebas
# (basically just the escape sequences), and gives you an example of
# what the input can look like if you use all the bells and whistles. :-)
#
# See `demolbl.bas' for a label-using version.


10 rem zmakebas demo

# tabs (as below) are fine (they're removed)
20 let init=	1000:\
   let header=	2000:\
   let udgdem=	3000:\
   let blockdem=4000

30 gosub init
40 gosub header
50 gosub udgdem
60 gosub blockdem
70 stop


# init

1000 for f=usr "a"+7 to usr "u"+7 step 8
1010 poke f,255
1020 next f
1030 let is48=1

# init all attrs just in case
1040 border 7:paper 7:ink 7:bright 0:flash 0:cls

# check for 48k speccy or 48k mode. This is a pretty nasty way to do
# it, but seems to be the most sane way (from Basic at least).
1050 print "\t"
1060 if screen$(0,0)="S" then let is48=0
1070 ink 0:print at 0,0;
1090 return


# header

2000 print ink 5;"\..\..\..\..\..\..\..\..\..\..\..\..\..\..\..\..";\
	         "\..\..\..\..\..\..\..\..\..\..\..\..\..\..\.."
2010 print paper 5;"  Non-ASCII chars in zmakebas  "
2020 print ink 5;"\''\''\''\''\''\''\''\''\''\''\''\''\''\''\''\''";\
	         "\''\''\''\''\''\''\''\''\''\''\''\''\''\''\''"
2030 print
2040 return


# udgdem

3000 print "Here are the UDGs:"''
3010 print ink 1;"\a\b\c\d\e\f\g\h\i\j\k\l\m\n\o\p\q\r\s";
3020 if is48 then print ink 1;"\t\u";
3030 print ''"(They should be underlined.)"
3040 return


# blockdem

#                   01234567890123456789012345678901
4000 print at 11,0;"The block graphics, first as"'\
		   "listed by a for..next loop, then"'\
		   "via zmakebas's escape sequences:"
4010 ink 7
4020 print at 15,0;
4030 for f=128 to 143:print chr$(f);" ";:next f:print ''
4040 print at 17,0;"\   \ ' \'  \'' \ . \ : \'. \': ";\
		   "\.  \.' \:  \:' \.. \.: \:. \::"
# draw boxes around them to make it look less confusing
4050 ink 1
4060 for y=0 to 1
4070 for x=0 to 15
4080 plot x*16,55-y*16:draw 7,0:draw 0,-7:draw -7,0:draw 0,7
4090 next x
4100 next y
4110 ink 0
4120 print at 20,0;"And finally here's the copyright"'\
		   "symbol (";ink 1;"\*";ink 0;\
		   ") and pound sign (";ink 1;"`";ink 0;")."
4130 return
